### ISSUE TITLE
introduce logging

### ISSUE THREAD

**Issue by JanSchulz** [IRkernel/IRkernel#119]
Currently we have some errors which are hard to debug. It also would be nice to tell users a switch and then get some logging output delivered in the console or a file.

from some short searching, I found https://github.com/smbache/loggr

**Comment by JanSchulz** [comment]
I also added a issue to build a debugging lag into the kernel commandline: https://github.com/jupyter/jupyter_client/issues/11

**Comment by flying-sheep** [comment]
also:

* https://github.com/zatonovo/futile.logger
* https://github.com/mfrasca/r-logging

**Comment by flying-sheep** [comment]
what i like in a logger are

1. namespaces (kernel log? executor log?) via logging objects or somehow automatically
2. wrapping/catching native facilities (in R: `warning` und `message`)
3. also if/when we go in CRAN, we’ll want a package on CRAN.

---

loggr und logr seem to do 2., while all other do 1. log4r, futile.logger and logging do 3.

**Comment by flying-sheep** [comment]
so there’s still no package that has both flexible logging objects and integration with `message`, `warning`, and `stop`.

there’s two approaches:

1. `logr` uses the R internal machinery for its stuff and requires the [`with_logging`](https://github.com/cubranic/logr/blob/b003a45c16e08fb9e4e9d1fbb223df570fdbd20c/R/logging.R#L52-L76) wrapper to reroute those to a file.
2. `loggr` pokes around in the interna and [makes *them* use *its* machinery](https://github.com/smbache/loggr/blob/master/R/set_loggr_hook.R).

and still, both don’t use multilevel logging/logging objects.

**Comment by janschulz** [comment]
> wrapping/catching native facilities (in R: warning und message)

Don't we anyway have to do our own catching? At least in the evaluate call and probably also in the kernel itself, so that we don't break on bad inputs...

**Comment by flying-sheep** [comment]
nonono, not `stop` catching. i meant that code which relies on `message` and `warning` interacts with the logging system and not stupidly prints to stderr.

**Comment by janschulz** [comment]
> nonono, not stop catching. i meant that code which relies on message and warning interacts with the logging system and not stupidly prints to stderr

We would still need to catch these stuff in the user supplied code to get this to the frontend... So this helps us only with our own code in the kernel/repr/irdisplay and maybe evaluate, which AFAIK all do not use message/warning?

**Comment by flying-sheep** [comment]
display and repr use nothing (both pretty slim and testable, no infinite loops, …)

and you’re right about user code: we already have `handle_message` and `handle_warning`, which we can use to forward stuff to the kernel log, right?

**Comment by janschulz** [comment]
> BUT if there’s some warning thrown by our machinery’s use of builtin R functions, it’ll only end up in the kernel’s stdout (right?)

Yes, I think so.

A big nice "catch everything" around the main `kernel$run()` would be nice... :-) So that unexpected errors get logged (and then the kernel quits to be restarted by the notebook server).

**Comment by flying-sheep** [comment]
yup. or better just within `kernel$run()`, i.e. replace it with `withConditionHandlers(.self$actually_run(), warning = log_warn, message = log_info)` or so

**Comment by janschulz** [comment]
There is also the kernel nanny enhancement in jupyter, which is relevant: https://github.com/jupyter/enhancement-proposals/pull/14
